Implement frontend services for dynamic feature gating

MegrezZhu opened the pull request:
## Overview

[Design Doc of Dynamic Feature Gating](https://docs.google.com/document/d/1FjbG2Cb0OnDVuE36jzux2Sn0sybMQozkk3hjkfeMoC0/edit?usp=sharing)

1. This PR fixes or fixes part of N/A.
2. This PR does the following:
    * added related frontend services for feature gating (and corresponding tests)
        * backend api services for getting feature flag values and updating feature flag settings
        * feature gating service that loads and caches feature flag results

**Note**: this pull request only adds the mentioned service but they are not enabled in this PR, there will be another PR to do this.

/cc @BenHenning @vinitamurthi

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

MegrezZhu committed:
Add storage models

MegrezZhu committed:
Merge branch 'develop' into dynamic-feature-gating-backend-dev

MegrezZhu committed:
Rename models, add validation jobs and change domain object

MegrezZhu committed:
Merge branch 'develop' into dynamic-feature-gating-backend-dev

MegrezZhu committed:
Fix method naming

MegrezZhu committed:
Fix import error

MegrezZhu committed:
Fix test issues

MegrezZhu committed:
Improve coverage, fix linting errors, add codeowner

MegrezZhu committed:
Remove debug statements

MegrezZhu committed:
Add code owner entries

MegrezZhu committed:
Change validator id regex

MegrezZhu committed:
Add domain objects

MegrezZhu committed:
Add docstrings

MegrezZhu committed:
Improve test coverage

MegrezZhu committed:
Added requested changes

MegrezZhu committed:
Fix test errors

MegrezZhu committed:
Merge branch 'storage-dev' into domain-dev

MegrezZhu committed:
Fix test errors

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Change code owners

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Add missing docstrings, fix test errors

MegrezZhu committed:
Add service methods

MegrezZhu committed:
Add docstrings

MegrezZhu committed:
Improve test

MegrezZhu committed:
Add schema_version, improve docstrings

MegrezZhu committed:
Merge branch 'develop' into storage-dev

MegrezZhu committed:
Merge branch 'storage-dev' into domain-dev

MegrezZhu committed:
Styling

MegrezZhu committed:
Fix audit job tests

MegrezZhu committed:
Merge branch 'storage-dev' into domain-dev

MegrezZhu committed:
Add schema version

MegrezZhu committed:
Merge branch 'domain-dev' into service-dev

MegrezZhu committed:
Add method, fix test errors

MegrezZhu committed:
Add necessary imports

MegrezZhu committed:
Styling, fix codeowners

MegrezZhu committed:
Styling

MegrezZhu committed:
Add controllers

MegrezZhu committed:
Merge branch 'storage-dev' into domain-dev

MegrezZhu committed:
Add method for feature, add dummy instance for demo

MegrezZhu committed:
Merge branch 'domain-dev' into controller-dev

MegrezZhu committed:
Add test

MegrezZhu committed:
Merge branch 'develop' into domain-dev

MegrezZhu committed:
Merge branch 'domain-dev' into service-dev

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Fix linting errors on keyword args

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Refactor filter to use (op, value) schema

MegrezZhu committed:
Add requested change

MegrezZhu committed:
Add browser type check

MegrezZhu committed:
Change docstring

MegrezZhu committed:
Add rquested changes

MegrezZhu committed:
Change test class name

MegrezZhu committed:
Move Registry

MegrezZhu committed:
Add front end domain objects

MegrezZhu committed:
Add front end domain objects

MegrezZhu committed:
Add code owners

MegrezZhu committed:
Add code owners

MegrezZhu committed:
Merge branch 'develop' into fe-domain-dev

MegrezZhu committed:
Fix linting error, add result object

MegrezZhu committed:
Fix linting error, add result object

MegrezZhu committed:
Remove debugging statements

MegrezZhu committed:
Remove debugging statements

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Improve test case names

MegrezZhu committed:
Fix docstring

MegrezZhu committed:
Fix test error

MegrezZhu committed:
Rename result object

MegrezZhu committed:
Rename result object

MegrezZhu committed:
Add client context object, fix test error

MegrezZhu committed:
Add client context object, fix test error

MegrezZhu committed:
Improve test naming, add requested changes

MegrezZhu committed:
Rename helper method

MegrezZhu committed:
Remove metadata object, add comment for tests

MegrezZhu committed:
Add app_version_flavor filter

MegrezZhu committed:
Change none value handling

MegrezZhu committed:
Merge branch 'domain-dev' into service-dev

MegrezZhu committed:
Sync interface, refactor and improve test

MegrezZhu committed:
Edit codeowners

MegrezZhu committed:
Fix linter error

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Sync interfaces, fix codeowner

MegrezZhu committed:
Fix docstrings

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Merge branch 'controller-dev' into fe-domain-dev

MegrezZhu committed:
Remove metadata object

MegrezZhu committed:
Remove metadata object

MegrezZhu committed:
Merge branch 'develop' into domain-dev

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Add comparison support to flavor filters

MegrezZhu committed:
Merge branch 'domain-dev' into service-dev

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Merge branch 'controller-dev' into fe-domain-dev

MegrezZhu committed:
Fix bug with None version

MegrezZhu committed:
Merge branch 'domain' into service-dev

MegrezZhu committed:
Change expected payload format

MegrezZhu committed:
Merge branch 'controller' into fe-domain-dev

MegrezZhu committed:
Add frontend services

MegrezZhu committed:
Improve test and docstrings

MegrezZhu committed:
Fix import sorting

MegrezZhu committed:
Add test, remove fdescribe

MegrezZhu committed:
Fix flavor ordering, fix docstring

MegrezZhu committed:
Add requested change

MegrezZhu committed:
Merge branch 'domain-dev' into service-dev

MegrezZhu committed:
Sync interface

MegrezZhu committed:
Fix redirect bug, add codeowner, fix fileoverview

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Merge branch 'controller-dev' into fe-domain-dev

MegrezZhu committed:
Merge branch 'fe-domain' into fe-service-dev

MegrezZhu committed:
Merge branch 'develop' into service-dev

MegrezZhu committed:
Merge branch 'develop' into service-dev

MegrezZhu committed:
Merge branch 'develop' into controller

MegrezZhu committed:
Rename files, improve comment & test

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Add interface description

MegrezZhu committed:
Link todo with issue

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Delete validation

MegrezZhu committed:
Make properties static

MegrezZhu committed:
Add guard for initialization, add docstring

MegrezZhu committed:
Fix test

MegrezZhu committed:
Sync change

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Improve interface

MegrezZhu committed:
Rename files using new pattern

MegrezZhu committed:
Rename FeatureFlagResults -> FeatureStatusSummary

MegrezZhu committed:
Rename files & fix typo

MegrezZhu committed:
Fix spacing, improve docstring

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Rename files

MegrezZhu committed:
Sync interface

MegrezZhu committed:
Sync renaming

MegrezZhu committed:
Fix naming

MegrezZhu committed:
Rename files

MegrezZhu committed:
Fix naming

MegrezZhu committed:
Merge branch 'develop' into service-dev

MegrezZhu committed:
Merge branch 'develop' into fe-domain-new-dev

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Improve error handling & add test for invalid rule update

MegrezZhu committed:
Merge branch 'controller-dev' into controller

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Rename files, sync changes

MegrezZhu committed:
Fix linting error

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Improve interface

MegrezZhu committed:
Fix type, fix codeowner, add docstring

MegrezZhu committed:
Merge branch 'develop' into service-dev

MegrezZhu committed:
Merge branch 'service-dev' into controller-dev

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Remove debug statements

MegrezZhu committed:
Sync interface changes, fix test

MegrezZhu committed:
Fix request bug

MegrezZhu committed:
Merge branch 'develop' into fe-service-dev

MegrezZhu committed:
Add missing test, fix test error, fix linting error

MegrezZhu committed:
Merge branch 'develop' into controller-dev

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

MegrezZhu committed:
Sync changes

BenHenning commented on core/controllers/base.py:
```diff
@@ -248,7 +248,12 @@ def dispatch(self):
-        if self.partially_logged_in:
+        # The 'platformfeaturehandler' requests should be exceptions here, if
+        # not, requests for feature flags when the user is partially logged in
+        # will cause their session cookie being deleted, leading to the
```
Why do such requests result in the session cookie being deleted? This exception doesn't seem ideal, and I don't understand why it's needed. We should try to avoid special casing pages like this if possible.

seanlip replied:
+1, I am concerned about this. Perhaps more thought is needed to find a way to keep things general at this layer.

MegrezZhu replied:
When a user hasn't completed the sign up process, it's marked as 'partially_logged_in' and redirected to the sign up page, any other request (the feature flag request especially) will be redirected to the `/logout` url which then deletes all cookies.
An alternative is to special case the signup page in `featureGatingService.initialize` instead of in `base.py`, so that in the `base.py` things are still general.

BenHenning replied:
Special casing initialize() does seem a lot cleaner to me, and more plug-and-play.

BenHenning commented on core/templates/base-components/oppia-root.directive.ts:
```diff
@@ -44,8 +44,10 @@ angular.module('oppia').directive('oppiaRoot', ['$translate',
+
             // The next line allows the transcluded content to start executing.
             $scope.initialized = true;
+            $scope.$apply();
```
Why is this needed now?

MegrezZhu replied:
The `$apply` is needed otherwise the page won't be loaded after I trigger `$scope.onInit` as the promise resolves. I'm not 100% sure but I think this might be neccessary if we change `$scope.initialized` in an asynchronous manner.

BenHenning replied:
@seanlip do you have any thoughts on why this $apply might be needed?

srijanreddy98 replied:
$apply is needed now because of the promise that has been introduced in this pr in oppia-angular-root (line no. 1723). It is not a problem in angular because angular uses zone.js, that monkey patches these promises to automatically trigger change detection when they resolve. However, angularjs doesn't use zonejs. So we have to manually trigger change detection by calling $scope.$apply(). Hope this clarifies why we need this $apply now.

BenHenning replied:
Aha, thanks for the explanation @srijanreddy98!

@MegrezZhu suggest adding a brief comment capturing the high-level aspects of that here to make sure the $apply stays in place.

MegrezZhu replied:
Done, added a comment to this line.

MegrezZhu committed:
Merge branch 'develop' into controller-dev

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Fix typo

MegrezZhu committed:
Merge branch 'develop' into controller-dev

MegrezZhu committed:
Fix errors after merging develop

BenHenning commented on core/templates/components/oppia-angular-root.component.ts:
```diff
@@ -1699,6 +1713,9 @@ private writtenTranslationsObjectFactory: WrittenTranslationsObjectFactory
     OppiaAngularRootComponent.writtenTranslationObjectFactory = this.writtenTranslationObjectFactory;
     OppiaAngularRootComponent.writtenTranslationsObjectFactory = this.writtenTranslationsObjectFactory;
     // This emit triggers ajs to start its app.
```
Probably want to remove this comment below to stay with the ``emit()`` line since that's what it's describing.

MegrezZhu replied:
Done, moved the comment.

BenHenning commented on core/templates/components/oppia-angular-root.component.ts:
```diff
@@ -1699,6 +1713,9 @@ private writtenTranslationsObjectFactory: WrittenTranslationsObjectFactory
     OppiaAngularRootComponent.writtenTranslationsObjectFactory = this.writtenTranslationsObjectFactory;
     // This emit triggers ajs to start its app.
-    this.initialized.emit();
+
```
Please add a comment explaining why this initialization is being done in this way.

BenHenning replied:
Also, please add tests for this case.

Finally, if platform parameter initialization fails does that mean that no pages will load? If so, that seems really problematic.

MegrezZhu replied:
Added comment for this line of code, also added test for the oppia-angular-root component.

When the initialization fails, this promise still resolves (because it's still initialized and the values are available, which are all False then). This is ensured by the top-level try..catch in `PlatformFeatureService._initialize`.

BenHenning replied:
Perfect--thanks!

BenHenning commented on core/templates/domain/platform_feature/feature-status-summary-object.factory.ts:
```diff
@@ -19,10 +19,40 @@
 import { Injectable } from '@angular/core';
 import { downgradeInjectable } from '@angular/upgrade/static';
 
+/**
+ * Names of all feature flags should be defined here, with format:
+ * FeatureName = 'feature_name', where the LHS is the feature name in
+ * PascalCase, and the RHS is in snake_case, which is the naming convention
+ * of features in the backend.
+ */
+export enum FeatureNames {
+  DummyFeature = 'dummy_feature',
+}
```
How do we make sure that this list matches the list in the backend? We should at least have some sort of test to ensure the two stay correct & in-sync to avoid inconsistencies.

MegrezZhu replied:
Added tests for this in `core/platform_feature_list_test`, that reads and parse the ts file to check the existence of feature names.
This solution seems suboptimal, I considered defining feature names in `constants.ts` or `feconf.py` but they all results to lost of member hinting as they are not ts files...

BenHenning replied:
Why would we lose typing if in constants.ts? Can we not define an enum in there & cross-validate the list with a Python test?

Not keen on the existing solution needing to parse a frontend file. I don't think we do that elsewhere, and the recommended pattern for sharing constants between the frontend & backend is constants.ts.

MegrezZhu replied:
I tried to define enum in `constants.ts` but there are some issues:

* inside `constants.ts` it uses `exports = {}` which makes it impossible to have another exported entry with `export enum {}`
* it's possible though to change it to `exports default {}` so that we can have other export entries, but this affects the whole base, tens of files needs to be changed and I'm not sure if simply changing the reference will have any issues in some js environment
  * even if I did so, and I created an enum for the names, I cannot add all the names as a list in the default exported json (to make it visible for the python end), because it requires runtime computation which makes it a ts/js file instead of a json, then in py it cannot be simply parsed.
* If I chose to simply add a name list as in the big constant object inside `constants.ts`. then I am not able to convert the list to enum because the list is considered to be a runtime variable, not a static constant.

It might still be possible to do so, but I think the work needed is not trivial...

BenHenning replied:
Okay, thanks for digging into it. Let's keep the current solution, and we can iterate on this in the future if the current solution actually introduces issues. The backend tests to keep things in sync are a good idea.

MegrezZhu commented:
@vojtechjelinek @seanlip @vinitamurthi @nithusha21 @DubeySandeep  @aks681 @bansalnitish @srijanreddy98 @jameesjohn 
PTAL at this pull request as reviewers/code owners, note that this PR is chained with previous PRs, so the changed files shown in this PR includes changes that are introduced in previous PRs, please just look at the files listed in the description of this pull request for now, and also feel free to take another pass once the previous PRs get merged and the delta is clean.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+
+  /**
+   * Inializes the PlatformFeatureService, it's guaranteed that it's only
+   * inialized once.
```
```suggestion
   * Inializes the PlatformFeatureService. This function guarantees that the service
   * is initialized only once for subsequent calls.
```

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+   * Returns the summary object of feature flags, which can be used to get the
+   * value of feature flags.
+   * Example:
+   *   platformFeatureService.featureSummary.DummyFeature.isEnabled === true.
```
```suggestion
   * value of feature flags.
   *
   * Example:
   *   platformFeatureService.featureSummary.DummyFeature.isEnabled === true.
```

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+   * @returns {FeatureSummaryDict} - Summary object of feature flags.
+   * @throws {Error} - If this method is called before inialization.
+   */
+  get featureSummary(): FeatureSummaryDict {
```
Here & elsewhere: prefer verb & verb phrases for naming, e.g. ``getFeatureSummary`` or ``isInitializedWithError``.

MegrezZhu replied:
Renamed to `isInitializedWithError`, but for `featureSummary`, since it's also a getter method, I think its naming should be the same as attributes, otherwise `service.getFeatureSummary.DummyFeature` seems a bit weird to me.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+   * @returns {string} - The name of the browser that is being used.
+   */
+  detectBrowserType(): string {
+    const ua = this.windowRef.nativeWindow.navigator.userAgent;
```
Prefer ``userAgent`` for naming.

MegrezZhu replied:
Done.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+  detectBrowserType(): string {
+    const ua = this.windowRef.nativeWindow.navigator.userAgent;
+
+    if (ua.includes('edg') || ua.includes('Edge')) {
```
Should the first string be ``'edge'``?

MegrezZhu replied:
It's `edg`, it's explained in [this doc of edge](https://docs.microsoft.com/en-us/microsoft-edge/web-platform/user-agent-string).

BenHenning replied:
Suggest adding a comment with a link to that article to provide context since others might assume this is a typo, too.

Also, I suggest doing the Edge check last. In terms of likelihood of the browser name accidentally showing up in the user agent, 'edge' is a much more common word than the order browser names per random English.

MegrezZhu replied:
Done.

MegrezZhu replied:
Update: The ordering of the browser check has to be in the original way, because the user agent of Edge also includes `Chrome`.

BenHenning replied:
Ah, darn. Nice catch. Probably worth adding a brief comment here, then, that the ordering is important to avoid accidentally mixing up browsers.

MegrezZhu replied:
Done, added comment for the ordering.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+  }
+
+  /**
+   * Checks if there's any error, e.g. request timeout, during initialization.
```
How is this function meant to be used? I'm not sure I understand how it should be used for users of this class (outside of tests). If it's only meant for testing purposes, is there another way to observe whether initialization failed? E.g. maybe by checking the promise returned by the initialize function? We should avoid programming backdoors into APIs for testing purposes.

MegrezZhu replied:
This getter function is not used for now, I thought it would be better if we provide interfaces for others to inspect the state of the service.
For the returned promise, IMO it should always resolve even there are errors -- the service is still initialized.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+          .loadFeatureFlagsFromServer();
+        this.saveResults();
+      }
+    } catch (err) {
```
We should probably log this error, otherwise debugging failures is going to be hard.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+  }
+
+  /**
+   * Reads and parse feature flag results from the sessionStorage.
```
``Reads and parses``

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+  }
+
+  /**
+   * Validates the result saved in sessionStorage.
```
Please define what 'valid' means in this context in the documentation.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+  private static SESSION_STORAGE_KEY = 'SAVED_FEATURE_FLAGS';
+  private static SESSION_STORAGE_CACHE_TTL = 12 * 3600 * 1000; // 12 hours
+
+  private static COOKIE_NAME_FOR_SESSION_ID = 'SACSID';
+  private static COOKIE_NAME_FOR_SESSION_ID_IN_DEV = 'dev_appserver_login';
```
Does Oppia define these? How can we be sure that these won't change?

MegrezZhu replied:
It's actually defined by app engine:
https://github.com/oppia/oppia/blob/1033cb1b1789580da1985f494ce0eaf480c92c04/core/controllers/base.py#L52-L69

Also as Oppia is enforcing HTTPS, I omited `ACSID`` here.

BenHenning replied:
That's not entirely ideal. Do your end to end tests cover both cookie values being correct? That is, if app engine changes the name of these cookies tomorrow, will we have tests covering that this doesn't work anymore?

@kevjumba will these cookie values change when we move away from webapp?

MegrezZhu replied:
I don't think it's feasible to have an e2e test for the cookie name in production, as in non-prod environment we only have the mock login interface (that creates 'dev_appserver_login').
Also, the `_clear_login_cookies` above faces the same potiential problem: it also takes the cookie names as a preset constant, if app engine changes the name of cookies tomorrow there aren't tests to check them (as far as I know).

BenHenning replied:
This is tricky, then. What would be the experienced behavior if the cookie names did change?

@seanlip do you have any thoughts on this? I'm not sure what the alternative is, but do we have any precedent for depending on the session ID elsewhere in Oppia's frontend?

MegrezZhu replied:
As for `_clear_login_cookies`, it may no longer be able to clear the login state.
As for `PlatformFeatureService`, the `.getSessionIdFromCookie` method would always return `null` and the user session related validation might not work as expected. But fortunately, since we don't have any user-specific filter at this moment, there should not be any difference if the parsed session id is always `null`.

BenHenning replied:
Ah, that's good to know. Thanks! In that case, let's go with the current solution and we can explore making this more robust against app engine versions in the future.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+// limitations under the License.
+
+/**
+ * @fileoverview A service for retriving feature flags.
```
I think much more documentation is needed here. This should explain the caching policy (including both TTL & session ID management), consistency guarantees for checking feature flags, and a brief introduction on what feature flags actually are.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.spec.ts:
```diff
@@ -0,0 +1,351 @@
+      })
+    );
+
+    it('should request only once if there are more than one calls to ' +
```
nit: ``...there are more than one call to...``

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.spec.ts:
```diff
@@ -0,0 +1,351 @@
+      () => {
+        platformFeatureService = TestBed.get(PlatformFeatureService);
+        expect(
+          () => platformFeatureService.featureSummary.DummyFeature.isEnabled
```
From an API perspective, is it possible to get the syntax ``platformFeatureService.DummyFeature.isEnabled`` & skip the summary middle step? The current approach seems a bit wordy.

MegrezZhu replied:
As far as I know about typescript, we are not able to automatically add enum-named members to a class while preserving the type & hint.. Not sure if there's any walkaround though.

BenHenning commented on core/templates/domain/platform_feature/feature-status-summary-object.factory.ts:
```diff
@@ -72,6 +117,13 @@ export class FeatureStatusSummaryObjectFactory {
     return new FeatureStatusSummary(backendDict);
   }
+
+  createDefault(): FeatureStatusSummary {
```
Please add documentation.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/domain/platform_feature/feature-status-summary-object.factory.ts:
```diff
@@ -19,10 +19,40 @@
+class FeatureSummaryDictItem {
+  constructor(private getterFn: () => boolean) {}
+
+  get isEnabled(): boolean {
```
Please add documentation for this & the class.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/domain/platform_feature/platform-feature-backend-api.service.spec.ts:
```diff
@@ -0,0 +1,86 @@
+        expect(successHandler).toHaveBeenCalledWith(
+          featureStatusSummaryObjectFactory.createFromBackendDict(responseDict)
+        );
+        expect(failHandler).not.toHaveBeenCalled();
```
Suggest also adding a test for the case when the fail handler should be called when a failure is triggered. I know this is a trivial case from an implementation perspective, but it's an important behavior to ensure is correct from an API perspective.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/domain/platform_feature/platform-feature-admin-backend-api.service.spec.ts:
```diff
@@ -0,0 +1,76 @@
+      flushMicrotasks();
+
+      expect(successHandler).toHaveBeenCalled();
+      expect(failHandler).not.toHaveBeenCalled();
```
Similar to a later comment, suggest adding a test for the failure case as well.

BenHenning commented on core/templates/domain/platform_feature/platform-feature-admin-backend-api.service.spec.ts:
```diff
@@ -0,0 +1,76 @@
+    httpTestingController.verify();
+  });
+
+  it('should make a reqeust to update the feature flag rules',
```
Typo: ``request``

MegrezZhu replied:
Done.

BenHenning commented on core/templates/domain/platform_feature/platform-feature-domain.constants.ts:
```diff
@@ -0,0 +1,27 @@
+
+export class PlatformFeatureDomainConstants {
+  public static PLATFORM_FEATURE_HANDLER_URL =
+    '/platform_features_evaluation_handler';
```
Nit: constant name should match the handler name, e.g. ``PLATFORM_FEATURES_EVALUATION_HANDLER_URL``

MegrezZhu replied:
Done.

BenHenning commented on core/templates/domain/platform_feature/platform-feature-domain.constants.ts:
```diff
@@ -0,0 +1,27 @@
+  'pages/admin-page/admin-page.constants';
+
+export class PlatformFeatureDomainConstants {
+  public static PLATFORM_FEATURE_HANDLER_URL =
```
Suggest adding brief doc comments for each of these constants to provide context for what they are.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.spec.ts:
```diff
@@ -0,0 +1,351 @@
+  });
+
+  afterEach(() => {
+    clearStaticProperties();
```
Do we need to do this in both beforeEach and afterEach? Suggest picking one.

MegrezZhu replied:
Done, moved them into `beforeEach`

BenHenning commented on core/templates/services/platform-feature.service.spec.ts:
```diff
@@ -0,0 +1,351 @@
+      })
+    );
+
+    it('should load from sessionStorage if there\'s valid results.', fakeAsync(
```
Nit: ``...if there are valid results.``

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.spec.ts:
```diff
@@ -0,0 +1,351 @@
+          })
+        });
+
+        tick(60 * 1000);
```
Suggest adding a brief comment explaining why this wait is needed for the behavior being tested.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.spec.ts:
```diff
@@ -0,0 +1,351 @@
+          })
+        });
+
+        tick(13 * 3600 * 1000); // 13 hours later.
```
Suggest clarifying why 13 hours is significant here.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.spec.ts:
```diff
@@ -0,0 +1,351 @@
+    );
+
+    it(
+      'should load from server if sessionId of saved result does not match.',
```
Do you also need a test to verify that the server is not called if the session does match?

BenHenning replied:
Reminder to consider this comment.

MegrezZhu replied:
I think this is covered in the `should load from sessionStorage if there are valid results` case above, there is `expect(serverApi).not.haveBeenCalled`.

BenHenning reviewed (changes requested):
Thanks @MegrezZhu! Did a full pass on the PR. At a high-level, the approach makes sense. I have some API-specific suggestions, but most of my comments are nits.

I also think other folks more experienced with Oppia web definitely need to weigh in on the nuances of how we're hooking up feature flags, and the specifics of the consistency mechanism (especially session ID checking), but I don't currently see any major problems. @seanlip might you be the right person to assess this?

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+   * @returns {boolean} - True if the result is valid and can be directly used.
+   */
+  private validateSavedResults(item: FeatureFlagsCacheItem): boolean {
+    if (this.getCurrentTimestamp() - item.timestamp >
+        PlatformFeatureService.SESSION_STORAGE_CACHE_TTL) {
```
Suggest adding tests specifically to make sure only the outdated parameters are updated since this is an important behavior to ensure.

MegrezZhu replied:
Not sure if I understand this.
By 'only the outdate parameters', are you implying that there are other parameters that are not outdated? Since now all parameters have universal TTL, they should expire as a whole.

BenHenning replied:
Per the code and your design doc, my understanding was that we track individual parameters for being outdated. I recall a comment in your design doc indicating that only outdated feature flags are updated when they reach their TTL threshold. My suggestion is to ensure there's a test to verify this (e.g. if you add one feature at time X, advance the clock, add another feature Y, and advance the clock such that X is expired but Y is not, X should be updated).

MegrezZhu replied:
Ah that was the original design, but then we agreed to set a universal 12h TTL for the entire features result.

But this does remind me of the case where a new feature is added to the code of Web, the page is refreshed and then there will be a mismatch with the stored feature results -- added a new constraint in `validateSavedResults`.

seanlip commented:
> I also think other folks more experienced with Oppia web definitely need to weigh in on the nuances of how we're hooking up feature flags, and the specifics of the consistency mechanism (especially session ID checking), but I don't currently see any major problems. @seanlip might you be the right person to assess this?

@BenHenning I might be, but I'm not sure what exactly I'm supposed to be looking at (this PR is quite large and I only am codeowner of core/controllers/base.py atm). Could you help me narrow it down?

vinitamurthi commented on core/templates/domain/platform_feature/platform-feature-domain.constants.ts:
```diff
@@ -0,0 +1,27 @@
+export class PlatformFeatureDomainConstants {
+  public static PLATFORM_FEATURE_HANDLER_URL =
+    '/platform_features_evaluation_handler';
+  public static ADMIN_HANDLER_URL = AdminPageConstants.ADMIN_HANDLER_URL;
```
Any reason why we cant directly use the Admin page constants handler URL wherever these constants are used, instead of a creating a redirection to that?

MegrezZhu replied:
Ah that makes sense, used `AdminPageConstants.ADMIN_HANDLER_URL` instead.

vinitamurthi commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,285 @@
+   *
+   * @returns {string} - The name of the browser that is being used.
+   */
+  detectBrowserType(): string {
```
Do we not already have a service to do this? If not, can we make this a separate service instead?

MegrezZhu replied:
Ah yes there's `BrowserCheckerService`, so I moved the `detectBrowserType` to this service.

vinitamurthi reviewed (approved):
Overall LGTM, I've left a couple of comments but they are not blocking. Approving this to unblock you further!

MegrezZhu committed:
Add requested changes

vojtechjelinek reviewed (approved):
LGTM!

BenHenning commented:
@seanlip I was mostly interested in your take on base.py--thanks.

MegrezZhu commented:
De-assigned codeowners of files only changed in previous PRs.

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Merge branch 'develop' into fe-service-dev

MegrezZhu committed:
Add tests, remove frontend coverage exception

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Merge branch 'controller-dev' into fe-service-dev

jameesjohn reviewed (approved):
LGTM for code owner files.

BenHenning commented on core/templates/domain/platform_feature/feature-status-summary-object.factory.ts:
```diff
@@ -118,6 +134,13 @@ export class FeatureStatusSummaryObjectFactory {
   }
 
+  /**
+   * Creates a default FeatureStatusSummary object that all features are
```
```suggestion
   * Creates a default FeatureStatusSummary object such that all features are
```

MegrezZhu replied:
Done.

BenHenning commented on core/templates/domain/platform_feature/feature-status-summary-object.factory.ts:
```diff
@@ -45,9 +45,25 @@ export type FeatureSummaryDict = {
+ * feature flag, providing the '.isEnabled' interface to check the status
+ * of that feature flag.
+ */
 class FeatureSummaryDictItem {
```
Something about the name of this class is confusing to me. What about ``FeatureStatusChecker``? I don't like 'summary' since it's a highly overloaded term in the Oppia codebase, and this doesn't look like other summary objects that we have.

BenHenning replied:
In general, let's try to avoid 'summary' here & elsewhere if it's not entirely clear what's being summarized. Also, lets be consistent on 'platform feature' vs 'feature' vs 'feature flag' in this PR & subsequent PRs.

MegrezZhu replied:
Done, renamed it to `FeatureStatusChecker`, also `FeatureStatusCheckerItem`.
Also renamed the member in `PlatformFeatureService` from `featureSummary` to `status`.

BenHenning commented on core/templates/domain/platform_feature/platform-feature-domain.constants.ts:
```diff
@@ -16,12 +16,12 @@
+
+  // Action name for request to the admin handler that updates the rules
+  // of feature flags.
   public static UPDATE_FEATURE_FLAG_RULES_ACTION = 'update_feature_flag_rules';
```
If this is specific to the admin page, should it be part of admin constants, instead?

MegrezZhu replied:
Ideally, yes, but there isn't a `AdminApiConstant`, and the other action strings are just defined where they are being used.

BenHenning commented on core/templates/domain/utilities/browser-checker.service.ts:
```diff
@@ -61,6 +61,30 @@ export class BrowserCheckerService {
   }
+
+  /**
+   * Detects the type of browser from it's userAgent.
```
```suggestion
   * Detects the type of browser from its user agent.
```

Nit on wording.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/domain/utilities/browser-checker.service.ts:
```diff
@@ -61,6 +61,30 @@ export class BrowserCheckerService {
+  detectBrowserType(): string {
+    const userAgent = this.windowRef.nativeWindow.navigator.userAgent;
+
+    if (userAgent.includes('edg') || userAgent.includes('Edge')) {
```
I suggest adding a comment here mentioning that the returned browser names shouldn't be changed since they're directly used for filtering platform parameters.

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -13,7 +13,17 @@
+ *   - the cache TTL of 12 hours has been reached, or
+ *   - the current account is different than the account in use when the values
+ *     are loaded, i.e. a different session id is present in the cookies.
+ * In such cases, the values will be re-initialized, they may be changed.
```
```suggestion
 * In such cases, the values will be re-initialized and they may be changed.
```

MegrezZhu replied:
Done.

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -146,20 +152,31 @@ export class PlatformFeatureService {
+      // The user is 'partially logged-in' at the signup page, we need to skip
+      // the loading from server otherwise the request will have the cookies
+      // erased, leading to the 'Registration session expired' error.
+      if (this.urlService.getPathname() === '/signup') {
+        PlatformFeatureService._isSkipped = true;
+        PlatformFeatureService.featureStatusSummary = this
+          .featureStatusSummaryObjectFactory.createDefault();
+        return;
```
Will the features be correctly loaded before the user reaches a non-signup page?

BenHenning replied:
Also, is this scenario being tested?

MegrezZhu replied:
Yes, after signing up the user will be redirected to a new page, which is a page refresh so the service will be initialized again in a non-signup page.
I have written tests for the exception in signup page. As for the post-signup redirection flow, I think it's beyond the test of the service itself because it will be a new page, with new PlatformFeatureService. -- The correctness is guaranteed in the e2e test though, where I created several accounts, and they are working well.

BenHenning replied:
Thanks for clarifying!

BenHenning commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -146,20 +152,31 @@ export class PlatformFeatureService {
       if (item && this.validateSavedResults(item)) {
         PlatformFeatureService.featureStatusSummary = item.featureStatusSummary;
         this.saveResults();
```
Why resave if it was just loaded?

MegrezZhu replied:
To refresh the TTL.
This is implied by the scenario where the user is actively using Oppia, in such cases results should not expire.

BenHenning reviewed (changes requested):
Thanks @MegrezZhu! Took another pass. I think the documentation improvements are quite good--thanks for doing them!

MegrezZhu committed:
Add checks for new feature

DubeySandeep reviewed (approved):
Approving this PR as I don't own any files specific to this PR! 

[I've already approved the files which I codeowner in the parent PR #10021)

bansalnitish reviewed (approved):
LGTM for codeowner files!

srijanreddy98 commented on core/templates/domain/platform_feature/platform-feature-backend-api.service.ts:
```diff
@@ -0,0 +1,68 @@
+ * @fileoverview Service to get feature flags.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+
+import { PlatformFeatureDomainConstants } from
+  'domain/platform_feature/platform-feature-domain.constants';
+import {
+  FeatureStatusSummary,
+  FeatureStatusSummaryBackendDict,
+  FeatureStatusSummaryObjectFactory,
+} from 'domain/platform_feature/feature-status-summary-object.factory';
+import { ClientContext } from
+  'domain/platform_feature/client-context-object.factory';
```
import order:

1. Angular imports
2. Third-party imports
3. Application imports.

Within these sub-categories: sort lexicographically based on module (i.e. where it is imported from, not what is being imported).
For example:
#### Right way (because 'B' < 'D')
```
import { C } from 'B';
import { A } from 'D'; 
```

#### Wrong way
```
import { A } from 'D';
import { C } from 'B'; 
```
```suggestion
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { downgradeInjectable } from '@angular/upgrade/static';

import { ClientContext } from
  'domain/platform_feature/client-context-object.factory';
import {
  FeatureStatusSummary,
  FeatureStatusSummaryBackendDict,
  FeatureStatusSummaryObjectFactory,
} from 'domain/platform_feature/feature-status-summary-object.factory';
import { PlatformFeatureDomainConstants } from
  'domain/platform_feature/platform-feature-domain.constants';
```

MegrezZhu replied:
Done, thanks for the notice!

srijanreddy98 commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,305 @@
+  private static COOKIE_NAME_FOR_SESSION_ID_IN_DEV = 'dev_appserver_login';
+
+  // TODO(#9154): Remove static when migration is complete.
+  static featureStatusSummary: FeatureStatusSummary = null;
```
@MegrezZhu static is no longer required. Apologies for the inconvenience as I didn't close the issue yet.

MegrezZhu replied:
Just to confirm, does that means there won't be multiple instances of the service even if it's being used in both Angular & AngularJS?

srijanreddy98 replied:
Yes! That happened because we used the service instance from UpgradedServices and not the one angular provided. This is not the case anymore. We have a way to use the service instance provided by angular without running into the angular injector error.

MegrezZhu replied:
Not sure if I understand this change correctly, but when I add logging statement in the constructor of the `PlatformFeatureService`, I can see from the messages that this service is instantiated twice...
Since this service contains important information (the feature status) I want to ensure it's the same in the page, and that's why I'm using static. Given that it's still possible to be inistantiated more than once, I prefer to keep the members static -- I can remove the TODO comments though given since this issue has been closed.

srijanreddy98 replied:
It is happening twice because you have listed it in upgraded-services.ts file. Remove it from there and you will find it is instantiated only once.

srijanreddy98 commented on core/templates/services/platform-feature.service.ts:
```diff
@@ -0,0 +1,305 @@
+ *
+ * Once the initialization is done, the value of each feature flag is guaranteed
+ * to be constant within the page.
+ * The values are also cached in SessionStorage, so that even after page
```
Should we use localStorage? What happens when the user opens a new tab. Will the session storage persist? As far as I know, sessionStorage is bound not only to the origin but also to the browser tab. And there are a lot of links in our app that open a new tab.

MegrezZhu replied:
That is expected, per the design doc of this project, we guarantees the features in the same tab to be stable within the tab session (and 12h at most). When there's a new tab, as pulling new features does not break the experience, we will want to pull the up-to-date values.

BenHenning replied:
Given that this came up in review, I suggest making this super explicit in the documentation so that others understand the context as well.

MegrezZhu replied:
Added descriptions at the top comments.
Also added section 7.1 in the design doc for the reason why session storage is chosen over local storage.

srijanreddy98 replied:
Ah okay! Thanks for the explanation.

srijanreddy98 reviewed (changes requested):
@MegrezZhu I have some minor nits and a question. Please take a look!

nithusha21 reviewed (approved):
Nothing to add to existing comments. LGTM for codeowner files!

MegrezZhu committed:
Add requested changes

MegrezZhu committed:
Merge branch 'develop' into fe-service-dev

BenHenning commented on core/templates/domain/platform_feature/feature-status-summary-object.factory.ts:
```diff
@@ -34,25 +34,25 @@ export interface FeatureStatusSummaryBackendDict {
 
 /**
- * Summary of feature flags, which are keyed on their names defined in
+ * Statuc checker of feature flags, which are keyed on their names defined in
```
```suggestion
 * Status checker of feature flags, which are keyed on their names defined in
```

MegrezZhu replied:
Done.

MegrezZhu committed:
Fix browser check

MegrezZhu commented:
@srijanreddy98 PTAL :)

BenHenning reviewed (changes requested):
Thanks @MegrezZhu! Mostly nits left, but I'd like to get some thoughts from @seanlip or someone else on the cookie piece in particular before merging.

MegrezZhu committed:
Add requested comments

MegrezZhu committed:
Merge branch 'develop' into fe-service-dev

srijanreddy98 reviewed (changes requested):
@MegrezZhu -- LGTM! One last thing left. Can you remove the static declaration of class variables? https://github.com/oppia/oppia/pull/10185/files?file-filters%5B%5D=.ts&file-filters%5B%5D=No+extension#r474432444

MegrezZhu commented:
@srijanreddy98 thanks for the reply, but I still have some concerns about the static members, PTAL at my reply in [this thread above](https://github.com/oppia/oppia/pull/10185#discussion_r474732548).

MegrezZhu committed:
Merge branch 'develop' into fe-service-dev

BenHenning reviewed (approved):
Thanks @MegrezZhu! This looks really good.

srijanreddy98 reviewed (approved):
I am ok with static right now, can you maybe try it with the refactor later on.

Showtim3 commented:
Hey @MegrezZhu Don't you want a review from @seanlip  and @aks681?, if that's the case, make sure to assign them to this PR.(Sorry if i am mistaken.)

MegrezZhu committed:
Fix protractor hang in exp player page

MegrezZhu committed:
Merge branch 'develop' into fe-service

MegrezZhu committed:
Revert hook

MegrezZhu committed:
Merge branch 'develop' into fe-service-dev

MegrezZhu commented:
> Hey @MegrezZhu Don't you want a review from @seanlip and @aks681?, if that's the case, make sure to assign them to this PR.(Sorry if i am mistaken.)

Thanks for the noticing! I have updated the reviewers list. :)

MegrezZhu commented:
@BenHenning PTAL at the latest changes in this PR. 

All tests (except the newly added End-to-End Tests that also fails in other PRs) has passed, I think this PR is ready for merging, but since I changed the purpose of this PR and removed the init hook, you might want to take a look at those changes.

BenHenning reviewed (approved):
Thanks @MegrezZhu! The revision of initialization LGTM. Please make sure the PR description includes a note that this doesn’t actually enable the service and that will be done in a later PR. 

Also, please add a TODO + an issue to track initializing the service.
